hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpc/nio/NioAsyncSocket.java

```diff
@@ -332,29 +423,52 @@ public void activate(Reactor reactor) {
             throw new IllegalStateException("Can't activate " + this + ": readhandler isn't set");
         }
 
-        this.reactor = checkInstanceOf(NioReactor.class, reactor, "reactor");
+        this.reactor = reactor;
+        this.selector = reactor.selector;
         this.eventloopThread = reactor.eventloopThread();
         this.unflushedBufs = new MpmcArrayQueue<>(unflushedBufsCapacity);
+        this.receiveBuffer = ByteBuffer.allocateDirect(getReceiveBufferSize());
+        try {
+            key = socketChannel.register(selector, clientSide ? 0 : OP_READ, handler);
+        } catch (ClosedChannelException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
 
-        // todo: if something fails within this task, you just get a logged exception and that is it.
-        boolean offered = reactor.offer(() -> {
-            try {
-                selector = this.reactor.selector;
-                receiveBuffer = ByteBuffer.allocateDirect(getReceiveBufferSize());
-                if (!clientSide) {
-                    try {
-                        key = socketChannel.register(selector, OP_READ, reactorHandler);
-                    } catch (ClosedChannelException e) {
-                        throw new UncheckedIOException(e);
-                    }
-                }
-            } catch (Exception e) {
-                close();
+    @Override
+    public CompletableFuture<Void> connect(SocketAddress address) {
+        checkNotNull(address, "address");
+
+        if (logger.isInfoEnabled()) {
+            logger.info("Connecting to address:" + address);
+        }
+
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        if (Thread.currentThread() == eventloopThread) {
+            connect0(address, future);
+        } else {
+            if (reactor == null) {
```

Comment:
Should we have similar checks for the other methods that executes a runnable in reactor, just in case they can be called before activate?

Reply:
I don't want to litter the fast path with all kinds of checks.

Reply:
Then, why do we have it here? Can't we remove this as well?

Reply:
OK, that is fine for me 👍